### PR TITLE
Add support for authorization code authentication

### PR DESCRIPTION
**Pull request by vladbarosan**
Add support for OAuth2 Authorization Code flow for authentication. ( https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-protocols-oauth-code )

Thank you for your contribution to Go-AutoRest! We will triage and review it as soon as we can.

As part of submitting, please make sure you can make the following assertions:
 - [x] I've tested my changes, adding unit tests if applicable.
 - [ ] I've added Apache 2.0 Headers to the top of any new source files.
 - [ ] I'm submitting this PR to the `dev` branch, except in the case of urgent bug fixes warranting their own release.
 - [ ] If I'm targeting `master`, I've updated [CHANGELOG.md](https://github.com/Azure/go-autorest/blob/master/CHANGELOG.md) to address the changes I'm making.

### CODE REVIEW

**Comment by vladbarosan** [comment]
Addresses #149

File: autorest/adal/token.go [hunk 1]
```diff
@@ -384,6 +402,41 @@ func NewServicePrincipalTokenFromUsernamePassword(oauthConfig OAuthConfig, clien
 	)
 }
 
+// NewServicePrincipalTokenFromAuthorizationCode creates a ServicePrincipalToken from the
+func NewServicePrincipalTokenFromAuthorizationCode(oauthConfig OAuthConfig, clientID string, clientSecret string, authorizationCode string, redirectURI string, resource string, callbacks ...TokenRefreshCallback) (*ServicePrincipalToken, error) {
```
Comment (jhendrixMSFT) [inline autorest/adal/token.go]: How is the authorization code obtained?

Comment (vladbarosan) [reply autorest/adal/token.go]: It's returned by the credential provider ( e.g. AAD ).

Simple flow is kind of like this:
1. User calls `/auth` endpoint of web app.
2. Web app redirects to AAD and among others includes  the `redirectURI` ( which must match with the `redirectURI` configured on the AAD app that is used).
3. User auths to AAD and gets redirected to the `redirectURI` with the auth code in the request.

Comment (vladbarosan) [reply autorest/adal/token.go]: We will need to add a sample for this.

Comment (vladbarosan) [reply autorest/adal/token.go]: Azure-Samples/azure-sdk-for-go-samples#13

**Comment by vladbarosan** [comment]
Usage example : https://gist.github.com/vladbarosan/74019046b8e9c9a49ff67d6634e7736d

**Comment by vladbarosan** [comment]
Adding Python similar sample https://github.com/AzureAD/azure-activedirectory-library-for-python/blob/dev/sample/website_sample.py

**Pull request merged**